### DISCUSSION TITLE
Provide guidance on how to fetch Envoy metrics

### DISCUSSION THREAD

**Discussion by tomkerkhove** [higress-group/higress#289] [Q&A] [answered]
I'm curious to learn how end-users to pull Envoy metrics given Higress is based on Envoy:
![image](https://user-images.githubusercontent.com/4345663/231408397-f2b22c7d-e439-4c7a-bb67-44ec8989dc80.png)

I see the Highress image is [based on Istio](https://github.com/alibaba/higress/blob/main/docker/Dockerfile.higress) but has Envoy packages. Can we just scrape Higress gateway directly?

**Comment by johnlanni** [comment]
Here is a Chinese document：https://higress.io/zh-cn/docs/user/prometheus
@CH3CHO will provide an English version of it soon.

**Reply by pravinpushkar** [reply]
is there any metrics api which can be called to fetch metrics and how to use that ? 
Like where and what port is the metrics server running and how to use that.

**Reply by tomkerkhove** [reply]
So if I get the documentation correctly, then:
- The console allows end-users to use embedded Grafana dashboard to consume Prometheus metrics
- Prometheus metrics are exposed by Higress through scraping endpoint

Can you elaborate on:
- What Prometheus metrics are exposed by Higress?
- If the Envoy API can be used directly? If not, is there a reason why we can't use existing Envoy tools?

**Reply by johnlanni** [reply]
> is there any metrics api which can be called to fetch metrics and how to use that ? Like where and what port is the metrics server running and how to use that.

You can find these annotations in higress-gateway pods below. 
```yaml
 prometheus.io/path: /stats/prometheus
 prometheus.io/port: "15020"
```
That's the path to the metrics api.

**Reply by johnlanni** [reply]
> So if I get the documentation correctly, then:
> 
> * The console allows end-users to use embedded Grafana dashboard to consume Prometheus metrics
> * Prometheus metrics are exposed by Higress through scraping endpoint
> 
> Can you elaborate on:
> 
> * What Prometheus metrics are exposed by Higress?
> * If the Envoy API can be used directly? If not, is there a reason why we can't use existing Envoy tools?

You can access `127.0.0.1:15020/stats/promethues` in higress-gateway pod and you will get all Prometheus metrics. This is the same as istio/proxy.

**Reply by tomkerkhove** [reply]
Wonderful, thank you! We might want to document this as well.

**Reply by CH3CHO** [reply]
> Here is a Chinese document：https://higress.io/zh-cn/docs/user/prometheus @CH3CHO will provide an English version of it soon.

I'm working on the English version now. It should be ready by the end of this week. And it is a how-to guide type of document. Regarding the tech details, we may have it written in other doc.

**Reply by pravinpushkar** [reply]
Thanks @CH3CHO @johnlanni

**Reply by CH3CHO** [reply]
@pravinpushkar @tomkerkhove Here it is. Please let me know if you have any questions.
https://higress.io/en-us/docs/user/prometheus/

**Reply by pravinpushkar** [reply]
I am running a golang application which repeatedly fetches the metrics from above endpoint(/stats/promethus). I have few questions here -
~1. How long does it take to report back the metrics like these - `envoy_http_downstream_rq_total, envoy_vhost_route_upstream_rq_total , envoy_http_downstream_rq_time`. I am running for almost several minutes and still it shows 0s for above 3 metrics.~
~2. My application is trying to fetch the metrics after every 1 min.~

**Edit: The above 2 doubts are resolved.**

3. Is there a way to get only the diff of these metrics ?


@CH3CHO @johnlanni

**Reply by CH3CHO** [reply]
Sorry, it's beyond my knowledge. Could @johnlanni @gengleilei take a look at the question above?

**Reply by tomkerkhove** [reply]
Any thoughts on this @johnlanni @gengleilei?

**Reply by johnlanni** [reply]
> Is there a way to get only the diff of these metrics ?

I don't quite understand what you mean by diff.

**Reply by tomkerkhove** [reply, marked as answer]
Reported metrics accumulate over time and were wondering if it would be possible to report amount of new requests for every time window that is reported.

That way, no diff needs to be calculated - Meaning keep state of previous metric and subtract new metric with previous
